providers/implementations/kem/hybrid_kem.c

```diff
@@ -0,0 +1,260 @@
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    HYBRID_PKEY *auth = (HYBRID_PKEY *)vauth;
+
+    return hybrid_kem_init(ctx, auth, params, &EVP_PKEY_decapsulate_init);
+}
+
+static int hybrid_kem_auth_init(HYBRID_PKEY_CTX *ctx,
+                                HYBRID_PKEY *key, HYBRID_PKEY *auth,
+                                const OSSL_PARAM params[],
+                                int (*initf)(EVP_PKEY_CTX *ctx,
+                                             EVP_PKEY *auth,
+                                             const OSSL_PARAM params[]))
+{
+    unsigned int i;
+
+    if (!ossl_prov_is_running())
+        return 0;
+
+    for (i = 0; i < ctx->info->num_algs; i++) {
+        EVP_PKEY_CTX_free(ctx->ctxs[i]);
+        ctx->ctxs[i] = EVP_PKEY_CTX_new_from_pkey(ctx->libctx,
+                                                  key->keys[i], key->propq);
+        if (ctx->ctxs[i] == NULL || !(*initf)(ctx->ctxs[i], auth->keys[i], params))
+            return 0;
+    }
+    return ossl_hybrid_set_ctx_params(ctx, params);
+}
+
+int ossl_hybrid_kem_auth_encapsulate_init(void *vctx, void *vkey,
+                                          void *vauthpriv,
+                                          const OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    HYBRID_PKEY *key = (HYBRID_PKEY *)vkey;
+    HYBRID_PKEY *authpriv = (HYBRID_PKEY *)vauthpriv;
+
+    return hybrid_kem_auth_init(ctx, key, authpriv, params,
+                                &EVP_PKEY_auth_encapsulate_init);
+}
+
+int ossl_hybrid_kem_auth_decapsulate_init(void *vctx, void *vkey,
+                                          void *vauthpub,
+                                          const OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    HYBRID_PKEY *key = (HYBRID_PKEY *)vkey;
+    HYBRID_PKEY *authpub = (HYBRID_PKEY *)vauthpub;
+
+    return hybrid_kem_auth_init(ctx, key, authpub, params,
+                                &EVP_PKEY_auth_decapsulate_init);
+}
+
+int ossl_hybrid_kem_encapsulate(void *vctx, unsigned char *enc, size_t *enclen,
+                                unsigned char *secret, size_t *secretlen)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    size_t a, b;
+    unsigned int i;
+
+    if (ctx->key_length == 0)
+        return 0;
+
+    if (enc == NULL) {
+        if (enclen != NULL)
+            *enclen = ctx->ciphertext_bytes;
+        if (secretlen != NULL)
+            *secretlen = ctx->shared_secret_bytes;
+        return enclen != NULL || secretlen != NULL;
+    }
+
+    /* Error checking */
+    if (*secretlen < ctx->shared_secret_bytes) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH, "*secretlen too small");
+        return 0;
+    }
+    if (*enclen < ctx->ciphertext_bytes) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH, "*enclen too small");
+        return 0;
+    }
+
+    /* Set outputs */
+    *enclen = ctx->ciphertext_bytes;
```

Comment:
Better done later/right before return 1 (after all constituent operations succeeded)?

Reply:
Setting the output lengths correctly on a failed return is something we do do elsewhere.
I'm not too fussed either way.